expression/builtin_string.go

```diff
@@ -2611,70 +2612,105 @@ func splitToSubN(s string, n int) []string {
 	return subs
 }
 
-type insertFuncFunctionClass struct {
+type insertFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *insertFuncFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
+func (c *insertFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinInsertFuncSig{newBaseBuiltinFunc(args, ctx)}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt, tpInt, tpString)
+	bf.tp.Flen = mysql.MaxBlobWidth
+	SetBinFlagOrBinStr(args[0].GetType(), bf.tp)
```

Comment:
why?

Reply:
this is a bug fix: for `INSERT(str,pos,len,newstr)`, if `str` is a binary string, we should treat it like a byte array, otherwise we should treat it like a code point array. see `builtinInsertBinarySig` and `builtinInsertSig` for more detail

Reply:
If we use args[0].GetType() after newBaseBuiltinFunc,
args[0] has already been wrapped with cast.

Reply:
when `arg[0]` is a binary string, it will not be wrapped with a cast.

 btw, I think wrapped cast functions should guarantee the correct type inference, and every function should do its type inference based on its children, no matter what expression the children is